rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,369 @@
+# Feature Name: Generic UI Navigation API
+
+## Summary
+
+A set of components and a new system to define navigation nodes for the ECS UI.
+
+## Motivation
+
+On top of making it easy to implement controller navigation of
+in-game UIs, it enables keyboard navigations and answers accessibility concerns
+such as the ones
+[raised in this issue](https://github.com/bevyengine/bevy/issues/254).
+
+## User-facing explanation
+
+### Terminology
+
+* A **`NavRequest`** is the sum of all possible ui-related user inputs. It's what
+  the algorithm takes as input, on top of the _navigation tree_ to figure out
+  the next focused element.
+* A **`Focusable`** is an Entity that can be navigated-to and highlighted (focused)
+  with a controller or keyboard
+* A **`NavFence`** adds "scope" to `Focusable`s. All `Focusable`s children of a
+  `NavFence` will be isolated from other `Focusable` and will require special
+  rules to access and leave.
+* The **navigation tree** is the logical relationship between all `NavFence`s.
+  `Focusables` are not really part of the tree, as you never break out of the
+  `NavFence` without a special `NavRequest`
+* There is a single `Focused` Entity, which is the currently highlighted/used
+  Entity.
+* There can be several **`Active`** elements, they represent the "path" through the
+  _navigation tree_ to the current `Focused` element (there isn't an exact 1
+  to 1 relationship, see following examples)
+* The navigation algorithm **resolves** a `NavRequest` by finding which `Entity`
+  to next focus to
+* The **active trail** is all the `Active` elements from the outermost
+  `NavFence` up to the innermost `NavFence` containing the `Focused` element.
+
+### Game developer interactions with the system
+
+We add a _navigation tree_ to the UI system, the tree will be used to enable
+key-based (and joystick controllable) navigation of UI, by allowing users to
+rotate between different UI elements.
+
+Currently, the developer must themselves write the relationship between
+various buttons in the menus and specify which button press leads to other menu
+items, etc. Or more simply, limit themselves to pointer-based navigation to
+avoid having to think about menu navigation.
+
+By including navigation-related components such as `Focusable` and `NavFence`,
+the developer can quickly get a controller-navigable menu up and running.
+
+We could also add the `Focusable` component to standard UI widget bundles such
+as `ButtonBundle` and a potential `TextField`. This way, the developer _doesn't
+even need to specify the `Focusable` elements_.
```

Comment:
@TheRawMeatball 
> This is super exciting, though I have concerns about having people manually build the navigation tree. I think all navigation info should be derived from the existing widget tree, at no extra cost to the developer.

It is completely possible to make the widgets navigable without extra cost to the developers, simply by adding the `Focusable` component as a field to the various relevant component bundles. I didn't think of it earlier because I was locked in the "this is a plugin" mindset.